PXC-4469: Implement CLONE method for SST

kamil-holubicki opened the pull request:
Implement CLONE method for SST
Work in progress

Tusamarco committed:
inserting clone scripts

Tusamarco committed:
Working version of clone
This version has a patch to go straight to ist so is not fully functional unless removing that bypass

Tusamarco committed:
First version that works SST and IST

Tusamarco committed:
Working copy of wsrep_clone with PXC
SSL is NOT  supported

Tusamarco committed:
adding needed GRANTS

Tusamarco committed:
adding MTR tests

Tusamarco committed:
commenting unneeded debug

Tusamarco committed:
fixing debug setting

Tusamarco committed:
Update mysql_system_users.sql

bug.
Wrong comma

Tusamarco committed:
INFO not DEBUG

Tusamarco committed:
trying to identify why is not reporting the clone status in the mtr test
fixing config missed wsrep allowed methods

Tusamarco committed:
trying to identify why is not reporting the clone status in the mtr test
fixing config missed wsrep allowed methods

Tusamarco committed:
trying to identify why is not reporting the clone status in the mtr test
fixing config missed wsrep allowed methods

Tusamarco committed:
Working version also on MTR. I still need to find a way to be sure netcat is not running

Tusamarco committed:
SST clone with mtr successful

Tusamarco committed:
pre release

Tusamarco committed:
fix mtr

Tusamarco committed:
fixing result

Tusamarco committed:
clean up and netcat removal

Tusamarco committed:
Adding some help description

Tusamarco committed:
Change needed by different NetCat implementation on Debian

Tusamarco committed:
Add the clone script to CMake to install them in the right location.
Need to ad extension or make script will not identify them correctly

Tusamarco committed:
fix in the message

Tusamarco committed:
adjusting value to 1 given NC RH implementation does not support 0

Tusamarco committed:
Merge branch 'percona:8.0' into PXC-4469-sst-clone

Tusamarco committed:
fix permission adding DELETE

Tusamarco committed:
Merge branch 'percona:8.0' into PXC-4469-sst-clone

Tusamarco committed:
Adding revoke on SST user to remove the role association before drop the user
Some edit/cleanup as suggested by Kamil

Tusamarco committed:
Few refactoring:
- removed the need to perform the restart to clean up the final instance.
- use wsrep-sst-receive-address to get IP and port definition
- use one port for Netcat and MySQL
- some code cleanup
- Add message suppression when reporting clone Status %. It will be reported only if different from previous value.

Tusamarco committed:
Forgot to change the Cmake list

Tusamarco committed:
Enable SSL
Few minor refactoring to accomodate comments on PR

Tusamarco committed:
Fixing a bug for Debian netcat implementation

Tusamarco committed:
removing orphan code.

Tusamarco committed:
- removing instruction about mysqlx
- add check on Donor message to ensure message is complete before continue

Tusamarco committed:
Merge branch 'percona:8.0' into PXC-4469-sst-clone

Tusamarco committed:
Remote user (#3)

* Add code in wsrep_sst.cc to handle the remote usr/pw pair and pass it to STDIN
- modify bash script accordingly

Tusamarco committed:
fixing some formatting and  variable  naming.

Tusamarco committed:
Another format set as before also if i find this wrong

Tusamarco committed:
Fix a problem related to role revoke call. we need to silently execute it.
Some minor cleanup

Tusamarco committed:
Fix a problem related to role revoke call. we need to silently execute it.
Some minor cleanup

Tusamarco committed:
Add wsrep_on=off to all DONOR sql to isolate the action to one single node
Add removing sst user at the end of the process
commenting the CALL to bypass the warning on role tables.

Tusamarco committed:
We must hardcode the user for the moment given we do not pass it to joiner atm

Tusamarco committed:
We must hardcode the user for the moment given we do not pass it to joiner atm

Tusamarco committed:
fix few thing in the test mtr

kamil-holubicki committed:
Merge pull request #1995 from Tusamarco/PXC-4469-sst-clone

Pxc 4469 sst clone

kamil-holubicki committed:
PXC-4620: Fix MTR tests that fail after changes needed for clone SST

https://perconadev.atlassian.net/browse/PXC-4620

fixed:
galera.galera_sst_clone - aligned with galera_sst_xtrabackup-v2 test

Missing privileges granted / privileges cleanup:
main.grant_dynamic_flush_upgrade
main.roles-upgrade
main.mysql_upgrade_slave_master_info
main.plugin_auth
main.mysql_57_inplace_upgrade
main.mysql_80_inplace_upgrade
main.component-upgrade
main.grant_dynamic

re-recorded:
perfschema.privilege_table_io
funcs_1.is_table_privileges
funcs_1.is_schema_privileges
main.transactional_acl_tables
main.ps_sys_upgrade

kamil-holubicki committed:
Merge branch '8.0' into PXC-4469

kamil-holubicki committed:
PXC-4620: Fix MTR tests that fail after changes needed for clone SST

https://perconadev.atlassian.net/browse/PXC-4620

Addressed review comments.

Cleanup granting privileges for mysql.pxc.internal.session.

kamil-holubicki committed:
PXC-4620: Fix MTR tests that fail after changes needed for clone SST

https://perconadev.atlassian.net/browse/PXC-4620

We have to exclude pxc users during the upgrade to apply original rules
for non-pxc users. This the extension of commit edcfa59e (PXC-3641).

kamil-holubicki committed:
Merge pull request #2022 from kamil-holubicki/PXC-4620

PXC-4620: Fix MTR tests that fail after changes needed for clone SST

Tusamarco committed:
Fix PXC-4607

kamil-holubicki committed:
PXC-4585: Clone SST on joiner waits indefinitely if the donor node is killed while SST.

https://perconadev.atlassian.net/browse/PXC-4585

Problem 1:
If the donor node is gone after SST process is started, the joiner waits
indefinitely.

Problem 2:
If the joiner is killed during the SST process, the clone instance
remains alive in background.

Solution:
Introduced sst progress monitor on Joiner side. After inititiating
clone transfer it monitors a few things:
1. If the clone instance is still alive (if clone transfer is finished)
2. If the parent process is still alive. Actually this mechanism is
redundant, because if partent process dies, it should terminate its
children.
3. If there is any data transfer progress. If it stalls for
sst-idle-timeout (default 120 sec), SST process is terminated.

If we detect that SST has to be terminated (timeout, parent process
died) the script exits and the cleanup is done in EXIT handler.

If the parent process is terminated, it should terminate its children
as well. It termination is handle in TERM handler.

The similar handling is for netcat process.

Additionally converted all double brackets [[ ... ]] into single
brackets for consistency.

Tusamarco committed:
Merge pull request #2026 from Tusamarco/PXC-4607-2

Pxc 4607 Fix bug PXC-4607 clone plugin is left installed on the joiner node

Tusamarco committed:
Fix for PXC-4609 clone sst user password is getting logged in the error log

Tusamarco committed:
just remove the variable name

Tusamarco committed:
Some cosmetic changes

Tusamarco committed:
Some cosmetic changes

Tusamarco committed:
Some cosmetic changes empty spaces at the end

Tusamarco committed:
lib order in import ¯\_(ツ)_/¯

Tusamarco committed:
Some cosmetic changes empty spaces at the end

Tusamarco committed:
Merge pull request #2029 from Tusamarco/PXC-4609-2

Fix for PXC-4609 clone sst user password is getting logged in the err…

kamil-holubicki committed:
Merge pull request #2027 from kamil-holubicki/PXC-4585

PXC-4585: Clone SST on joiner waits indefinitely if the donor node is killed while SST.

Tusamarco commented on sql/wsrep_sst.cc:
```diff
@@ -276,17 +276,25 @@ static char *my_fgets(char *buf, size_t buf_len, FILE *stream) {
 struct sst_thread_arg {
   const char *cmd;
   char **env;
+  const sst_auth& auth_container;
```
@kamil-holubicki I assume we should address these clang issue before merging right?
me or you?

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+#----------------------------------------------------------------------------
+# Config variables:
+# [sst]
+# netcat_port=4442
```
This may not be valid now

Tusamarco replied:
true will remove it

kamil-holubicki replied:
done

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+# NOTE for SSL usage
+# When using clone SSL certificates MUST be manually (or by automation) placed in a location on both servers DONOR/RECEIVER
+# this location better not be the data directory given it is manipulated by the clone process.
+# It is reccomended to explicitly set the ssl certificates in the my.cnf as:
```
```suggestion
# It is recommended to explicitly set the ssl certificates in the my.cnf as:
```

kamil-holubicki replied:
done

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    fi
+}
+
+wsrep_check_program()
```
Duplicated function. This function is already defined in `wsrep_sst_common`. Please remove it.

Tusamarco replied:
right

kamil-holubicki replied:
done

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    fi
+}
+
+wsrep_check_programs()
```
Duplicated function. This function is already defined in `wsrep_sst_common`. Please remove it.

Tusamarco replied:
right

kamil-holubicki replied:
done

venkatesh-prasad-v commented on sql/wsrep_sst.cc:
```diff
@@ -1578,6 +1618,34 @@ int wsrep_sst_donate(const std::string &msg, const wsrep::gtid &current_gtid,
+  if (remote_auth()) {
+    /* wsp::string is just a dynamically allocated char* underneath
+     * so we can safely do all that arithmetics */
+    const char* col= strchrnul(remote_auth(), ':');
```
Here and everywhere, the space around `=` is inconsistent. Please run clang-format before merging.
```suggestion
    const char* col = strchrnul(remote_auth(), ':');
```

kamil-holubicki replied:
clang formatting for the whole file fixed

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    WSREP_SST_OPT_REMOTE_JOINER_PSWD=$WSREP_SST_OPT_REMOTE_PSWD
+
+    echo "continue" # donor can resume updating data
+    #Dead code ;) 
+    #WSREP_SST_OPT_REMOTE_AUTH=$(echo $WSREP_SST_OPT_ADDR_LOCAL | cut -d '@' -f 1)
+    #WSREP_SST_OPT_REMOTE_HOST_WITH_PORT=$(echo $WSREP_SST_OPT_ADDR_LOCAL | cut -d '@' -f 2) 
```
Do we still need it?

kamil-holubicki replied:
Removed

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    exit $EINVAL
+fi
+
+wsrep_log_debug "-> SST PROCESS FINISHED for $WSREP_SST_OPT_ROLE $?"
```
To which command does the `$?` here refer to?

kamil-holubicki replied:
fixed

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+                break
+            fi
+            if [ "$DONOR_TIMEOUT_WAIT_JOINER_CLONE_INSTANCE" == "0" ]; then
+                    wsrep_log_info "TIMEOUT Waiting for the Joiner to setup MySQL clone instance"
+                    wsrep_log_error "TIMEOUT Waiting for the Joiner to setup MySQL clone instance $LINENO"
```
There is no need to log the same message once as info and once as an error.

kamil-holubicki replied:
removed info

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    eval WSREP_SST_CLONE=1 $CLONE_ENV $CLONE_BINARY $DEFAULT_OPTIONS \
+        --wsrep_provider=none --init_file="$CLONE_SQL" >> $CLONE_ERR &
+
+    wsrep_log_info "-> Initialization user returned $?"
```
This message can be a debug message
```
2025-02-11T07:45:23.933270Z 0 [Note] [MY-000000] [WSREP-SST] -> Initialization user returned 0
```

Tusamarco replied:
Ok

kamil-holubicki replied:
done

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+EOF
+
+    wsrep_log_info "Launching clone recipient daemon"
+    wsrep_log_info "-> using: $CLONE_ENV $CLONE_BINARY_SAFE $DEFAULT_OPTIONS "
```
This is how the message looks in the error log
```
2025-02-11T07:45:23.932518Z 0 [Note] [MY-000000] [WSREP-SST] -> using: LD_LIBRARY_PATH=/home/venki/work/pxc/clone_sst/bld/lib/plugin/: /home/venki/work/pxc/clone_sst/bld/runtime_output_directory/mysqld_safe       --defaults-file='/home/venki/work/pxc/clone_sst/bld/mysql-test/var/my.cnf'      --defaults-group-suffix='.2'      --binlog-rotate-encryption-master-key-at-startup=OFF      --datadir='/home/venki/work/pxc/clone_sst/bld/mysql-test/var/mysqld.2/data/'      --wsrep_on=0      --secure_file_priv=''      --socket=/home/venki/work/pxc/clone_sst/bld/mysql-test/var/mysqld.2/data//recover_clone.sock      --log_error='/home/venki/work/pxc/clone_sst/bld/mysql-test/var/mysqld.2/data//clone_sst.err'      --pid_file='/home/venki/work/pxc/clone_sst/bld/mysql-test/var/mysqld.2/data//clone_sst.pid'      --plugin-dir='/home/venki/work/pxc/clone_sst/bld/lib/plugin/'      --port='13009'
```

Can we instead log it as debug?

Tusamarco replied:
ok

kamil-holubicki replied:
done

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+
+    if [ -n "$WSREP_LOG_DEBUG" ]; then
+        GRANTS=`$MYSQL_ACLIENT -uroot -NB -e "show grants for '$CLONE_USER'@'localhost'"`
+        wsrep_log_debug "-> Clone user grants: $GRANTS"
```
Do we really need this in our logs with wsrep-debug=1?
```
2025-02-11T07:45:25.787850Z 0 [Note] [MY-000000] [WSREP-SST] (debug) -> Clone user grants: GRANT SHUTDOWN, SUPER, EXECUTE ON *.* TO `clone_sst`@`localhost`
2025-02-11T07:45:25.787890Z 0 [Note] [MY-000000] [WSREP-SST] GRANT CLONE_ADMIN,SYSTEM_VARIABLES_ADMIN ON *.* TO `clone_sst`@`localhost`
2025-02-11T07:45:25.787898Z 0 [Note] [MY-000000] [WSREP-SST] GRANT SELECT, INSERT, UPDATE ON `performance_schema`.* TO `clone_sst`@`localhost`
2025-02-11T07:45:25.787905Z 0 [Note] [MY-000000] [WSREP-SST] GRANT INSERT ON `mysql`.`plugin` TO `clone_sst`@`localhost`
```

Tusamarco replied:
for debug yes

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    # this node possible.
+    # We need to use a temporary empty data directory, because the
+    # actual datadir may already contain some wsrep-related files
+    wsrep_log_info "Initializing data directory at $tmp_datadir"
+
+    wsrep_log_info "INITIALIZE DB"
```
We are already printing that we are initializing data directory.
```suggestion
    wsrep_log_info "Initializing data directory at $tmp_datadir"
```

kamil-holubicki replied:
removed INITIALIZED DB

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+  return 0
+}
+
+wsrep_log_debug "-> In the wsrep_sst_clone "
```
This message does not add any value. Let's remove it.
```
2025-02-11T07:45:11.214842Z 0 [Note] [MY-000000] [WSREP-SST] (debug) -> WSREP_SST_OPT_ADDR_LOCAL: 127.0.0.1:13009
2025-02-11T07:45:11.214858Z 0 [Note] [MY-000000] [WSREP-SST] (debug) -> WSREP_SST_OPT_LPORT: 
2025-02-11T07:45:11.215282Z 0 [Note] [MY-000000] [WSREP-SST] (debug) -> In the wsrep_sst_clone 
2025-02-11T07:45:11.232045Z 0 [Note] [MY-000000] [WSREP-SST] (debug) -> WSREP_SST_OPT_PORT 13009
2025-02-11T07:45:11.232068Z 0 [Note] [MY-000000] [WSREP-SST] (debug) -> WSREP_SST_OPT_ADDR_LOCAL 127.0.0.1:13009
```

Tusamarco replied:
not convinced,  this was useful to understand what transformation was done in common and what in the script, and how they were interacting. 
I would keep it is a debug info at the end

venkatesh-prasad-v replied:
I meant to only remove the line that prints `In the wsrep_sst_clone`

kamil-holubicki replied:
Done. Added comment that this is the point where the script execution starts.

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    IFS=':' read -ra my_array <<< "$WSREP_SST_OPT_ADDR_LOCAL"
+
+    lenAr=${#my_array[@]}
+    wsrep_log_debug "-> lenarray $lenAr"
```
Again, there is no need to print it to the log even with debug

Tusamarco replied:
there is, given we may have different formats, and this help us to understand the number of elements passed using the `:` as separator. 
Again debug info and to be honest I think the more info you have in debug the better.
In one year time we all have completely forgot how this works.

kamil-holubicki replied:
as long as it is debug, we can leave it as it is

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+         CLEANUP_FILES=1
+         cleanup_joiner ||:
+         wsrep_log_debug "->SENDING MESSAGE: $RP_PURGED "
+         wsrep_log_debug "->out code: $? "
```
No need to print `$?`.

Tusamarco replied:
as above

kamil-holubicki replied:
I'm afraid that printing $? is not what was intended to print. It prints the exit code of the previous command, so exit code of `wsrep_log_debug`. Two lines below we write exit code of `echo`.
Removing both logs

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+         echo $RP_PURGED >&2
+         echo $RP_PURGED
+         wsrep_log_debug "->out code: $? "
+         sleep 1
+         exit 0
```
Why `sleep 1` before exit?

kamil-holubicki replied:
removed

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+       wsrep_log_debug "-> Invalid Recovery position. Exiting with error 61 (No data available )"
+     exit 61
+   else
+    wsrep_log_ERROR "Something failed in the cleanup"
```
```suggestion
    wsrep_log_error "Something failed in the cleanup"
```

kamil-holubicki replied:
fixed

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+     CLEANUP_FILES=0
+     wsrep_cleanup_progress_file
+     cleanup_joiner
+       wsrep_log_debug "-> Invalid Recovery position. Exiting with error 61 (No data available )"
```
1. Incorrect formatting. 
2. Hardcoded error no 61. Consider using something similar to EINVAL

Tusamarco replied:
Why? EINVALis invalid argument, while this the check is specific which is ENODATA 61 No data available, referring to the position information.

venkatesh-prasad-v replied:
I meant to use it like 
```
   readonly EINVAL=22                                                                                                                                                                                          
   readonly EPIPE=32                                                                                                                                                                                           
   readonly ETIMEDOUT=110 
+  readonly ENODATA=61
```

kamil-holubicki replied:
ENODATA introduced

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    [ "$SERVER_SSL_KEY" = "NULL" ] && SERVER_SSL_KEY=
+    [ "$SERVER_SSL_CA" = "NULL" ] && SERVER_SSL_CA=
+
+    if [ "$CLONE_SSL_CERT" == "NULL" -o "$CLONE_SSL_CERT" = "" ]
```
Please address Kamil's comments
https://github.com/percona/percona-xtradb-cluster/pull/2018/files#r1939078846
>Mixture of single `=` and double `==` in one if

Tusamarco replied:
If I recall correctly we agreed to do this later, when we will beautify the script a bit more. there are many other things that can be changed but not affecting the functionalities.
If we decide to change all now fine for me it will just take other time.

kamil-holubicki replied:
Let's address it in a separate PR

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+            WSREP_OFF="" # joiner does not have replication enabled
+        fi
+        $MYSQL_ACLIENT -e "${WSREP_OFF}INSTALL PLUGIN CLONE SONAME 'mysql_clone.so';"
+        if [ "$ROLE" == "donor" ]; then
```
Please address Kamil's comments https://github.com/percona/percona-xtradb-cluster/pull/2018/files#r1939063985
>kamil-holubicki 
>
>It is not consistent across the script how we do the comparison. Sometimes it is == and sometimes it it = >(just 4 lines above).
>= is posix compatible and should be used with single brackets [...]
>== is bash only compatible and should be used with double brackets [[...]]
>
>Would it be possible to have it consistent?

Tusamarco replied:
as above and duplicate

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+EOF
+
+    fi
+    if [ "$?" == "0" ]; then
```
To which command does the `$?` refer to here?

kamil-holubicki replied:
Reworked this part.

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    eval $CLONE_ENV $CLONE_BINARY $DEFAULT_OPTIONS $SKIP_NETWORKING_OPTIONS --wsrep_recover >> $CLONE_ERR 2>&1
+
+    RP="$(grep -a '\[WSREP\] Recovered position:' $CLONE_ERR || :)"
+    RP_PURGED=`echo $RP | sed 's/.*WSREP\]\ Recovered\ position://' | sed 's/^[ \t]*//'`
```
Error returned by shellcheck

SC2006: Use $(...) notation instead of legacy backticks \`...\`.

kamil-holubicki replied:
done

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+            wsrep_log_error "Timeout waiting for clone recipient daemon"
+            exit $ETIMEDOUT
+        fi
+        to_wait=$(( $to_wait - 1 ))
```
SC2004: $/${} is unnecessary on arithmetic variables.

kamil-holubicki replied:
fixed

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    CLONE_INSTANCE_PID=`cat $CLONE_PID_FILE`
+
+    if [ -n "$WSREP_LOG_DEBUG" ]; then
+        GRANTS=`$MYSQL_ACLIENT -uroot -NB -e "show grants for '$CLONE_USER'@'localhost'"`
```
SC2006: Use $(...) notation instead of legacy backticks \`...\`.

kamil-holubicki replied:
done

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    do
+        sleep 0.2
+    done
+    CLONE_INSTANCE_PID=`cat $CLONE_PID_FILE`
```
SC2006: Use $(...) notation instead of legacy backticks \`...\`.

kamil-holubicki replied:
done

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+
+    # Before starting let us be sure we remove Netcat given it is using same MySQL port
+    # check if there is another netcat process on the IP port we need and try to kill it.
+    NETCAT_KILL=`ps -xo pid,command|grep -e "nc -l -k $JOINER_CLONE_HOST $JOINER_CLONE_PORT"|grep -v "grep"|cut -d ' ' -f 1`
```
SC2006: Use $(...) notation instead of legacy backticks `...`.

Tusamarco replied:
that is one of the other thing we can do to beautify the script and that has being postponed

kamil-holubicki replied:
done

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    ##################################################################################################
+    # If we need to SST in any case we must remove the data, so let us do it here and be sure we work on a clean directory
+    wsrep_log_info "Cleaning Data directory $WSREP_SST_OPT_DATA"
+    rm -fr $WSREP_SST_OPT_DATA/*
```
SC2115: Use "${var:?}" to ensure this never expands to /*

kamil-holubicki replied:
Yeah, this is good one. Nothing better than rm -rf /*
:)

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+     --datadir='$WSREP_SST_OPT_DATA' \
+     --wsrep_on=0 \
+     --secure_file_priv='' \
+     --socket="$CLONE_SOCK" \
```
SC2027: The surrounding quotes actually unquote this. Remove or escape them.
```suggestion
     --socket='$CLONE_SOCK' \
```

kamil-holubicki replied:
fixed

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    CLONE_SOCK="$WSREP_SST_OPT_DATA/recover_clone.sock"
+    CLONE_X_SOCK="$WSREP_SST_OPT_DATA/cmysqlx.sock"
+
+    DONOR_IP=""
```
Unused variable
SC2034: DONOR_IP appears unused. Verify use (or export if used externally).

Tusamarco replied:
yes

kamil-holubicki replied:
Removed

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+
+    CLONE_ERR="$WSREP_SST_OPT_DATA/$MODULE.err"
+    CLONE_SQL="$WSREP_SST_OPT_DATA/$MODULE.sql"
+    IST_FILE="$WSREP_SST_OPT_DATA/IST_FILE.txt"
```
Unused variable
SC2034: IST_FILE appears unused. Verify use (or export if used externally).

Tusamarco replied:
yes

kamil-holubicki replied:
removed

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    wsrep_log_debug "-> MYSQL_ACLIENT: $MYSQL_ACLIENT"
+
+    # HERE We start the instance that will get the clone and pass the CLONE_SQL file to create user
+    eval WSREP_SST_CLONE=1 $CLONE_ENV $CLONE_BINARY $DEFAULT_OPTIONS \
```
Where is the env variable `WSREP_SST_CLONE=1` used? If it is needed, please add a comment saying why it is needed

kamil-holubicki replied:
Removed.

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+. $(dirname "$0")/wsrep_sst_common #_clone
+
+JOINER_TIMEOUT_WAIT_XST=$(parse_cnf sst joiner-timeout-xst "60")
+JOINER_TIMEOUT_WAIT_CLONE_INSTANCE=$(parse_cnf sst joiner-timeout-clone-instance "90")
```
Is the default value of 90 seconds enough for most of the cases?

Tusamarco replied:
so far so good. and is a parameter so it can be change ... probably need to add on top of the script so it will result clearer

kamil-holubicki replied:
Let's leave as it is for now

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+
+. $(dirname "$0")/wsrep_sst_common #_clone
+
+JOINER_TIMEOUT_WAIT_XST=$(parse_cnf sst joiner-timeout-xst "60")
```
What does XST mean? Can we rename it to something that is more readable?

kamil-holubicki replied:
Renamed to JOINER_TIMEOUT_WAIT_DONOR_MESSAGE and joiner-timeout-wait-donor-message

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+MYPID=$$
+
+if [ "$WSREP_SST_OPT_PARENT" == "" ]; then
+        PARENT_PID=`cat /proc/$MYPID/status | grep PPid`
```
SC2006: Use $(...) notation instead of legacy backticks `...`.

Tusamarco replied:
as above

kamil-holubicki replied:
done

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    trap sig_cleanup_joiner HUP PIPE INT TERM
+    trap cleanup_joiner EXIT
+
+    wsrep_check_programs grep
+    wsrep_check_programs ps
+    wsrep_check_programs find
```
Script does not exit if any of the command is not found

kamil-holubicki replied:
I think it exists. The script runs without set +e, so it exists any time the function returns non-zero

kamil-holubicki replied:
yup, there is `set -o errexit` at the top of the script

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+if test -z "$WSREP_SST_OPT_HOST"; then wsrep_log_error "HOST cannot be nil"; exit $EINVAL; fi
+
+# MySQL client does not seem to agree to [] around IPv6 addresses
+wsrep_check_programs sed
```
The script does not stop or exit if `sed` is not installed. Can we have something like
```
   896         wsrep_check_program sed                                                                                                                                
   897     |   if [[ $? -ne 0 ]]; then                                                                                                                                
   898     |   |   wsrep_log_error "******** FATAL ERROR *********************** "                                                                                    
   899     |   |   wsrep_log_error "sed was not found in PATH! Make sure you have it installed."                                                                      
   900     |   |   wsrep_log_error "* Line $LINENO"                                                                                                                   
   901     |   |   wsrep_log_error "******************************************** "                                                                                    
   902     |   |   exit 2                                                                                                                                             
   903     |   fi
```

kamil-holubicki replied:
I think it exists. The script runs without set +e, so it exists any time the function returns non-zero

kamil-holubicki replied:
set -o errexit at the top of the script.

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+        # We stay on hold now, waiting for the Joiner to expose the service
+        wsrep_log_info "-> WAIT for Joiner MySQL to be available nc -w 1 -i 1 $SST_HOST_STRIPPED $WSREP_SST_OPT_REMOTE_HOSTPORT"
+        while [ true ]; do
+        	NCPING=`nc -w 1 -i 1 $SST_HOST_STRIPPED $WSREP_SST_OPT_REMOTE_HOSTPORT 2> /dev/null` || :
```
I found this warning in the log.
```
2025-02-11T07:45:25.764167Z 0 [Note] [MY-000000] [WSREP-SST] /home/venki/work/pxc/clone_sst/bld/scripts/wsrep_sst_clone: line 650: warning: command substitution: ignored null byte in input
```

kamil-holubicki replied:
Filtered out null bytes possibly produced by netcat

kamil-holubicki committed:
cleanup

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+
+    wsrep_log_info "-> Initialization user returned $?"
+
+    if [ ! "$?" == "0" ]; then
```
Again here, `$?` checks the return status of the previous command `wsrep_log_info`

kamil-holubicki replied:
fixed

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+SST_HOST_STRIPPED=$(echo $WSREP_SST_OPT_HOST | sed 's/^\[//' | sed 's/\]$//')
+
+# Option to CREATE USER
+WITH_OPTION="WITH caching_sha2_password"
```
Looks like there is no need for this as we are anyways creating users with `WITH caching_sha2_password`
```
 cat << EOF > "$CLONE_SQL"
 SET SESSION sql_log_bin=OFF;
 CREATE USER '$CLONE_USER'@'%' IDENTIFIED $WITH_OPTION BY '$CLONE_PSWD';
  ...
 CREATE USER '$CLONE_USER'@'localhost' IDENTIFIED $WITH_OPTION BY '$CLONE_PSWD';
```

Tusamarco replied:
I would prefer to have it in case someone change the defaults as I have seen quite often

kamil-holubicki replied:
Let's leave it as it is

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+ SET SESSION sql_log_bin=OFF;
+ CREATE USER '$CLONE_USER'@'%' IDENTIFIED $WITH_OPTION BY '$CLONE_PSWD';
+ GRANT CLONE_ADMIN, SYSTEM_VARIABLES_ADMIN, SUPER, SHUTDOWN, EXECUTE ON *.* to '$CLONE_USER'@'%';
+ GRANT INSERT ON mysql.plugin to '$CLONE_USER';
+ GRANT SELECT,UPDATE,INSERT ON performance_schema.* TO '$CLONE_USER';
```
Probably we need the host `%` as well
```suggestion
 GRANT INSERT ON mysql.plugin to '$CLONE_USER'@'%';
 GRANT SELECT,UPDATE,INSERT ON performance_schema.* TO '$CLONE_USER'@'%';
```

Tusamarco replied:
if omitted it defaulted to '%' anyhow. We can add it to be consistent and verbose, will not change the behaviour.

kamil-holubicki replied:
done

kamil-holubicki committed:
Addressed review comments.

kamil-holubicki commented:
There is a lot of backtick vs $(...) issues in the script. They are inherited for the original script. My proposal is not to touch them right now.

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+    "
+
+    # Define USER and PW
+    CLONE_PSWD=`wsrep_gen_secret`"$PSWD_POLICY"
+    CLONE_USER="clone_sst"
```
If the SST fails, then the CLONE_PSWD is exposed in the error log as 

```
2025-02-12T13:47:00.653175Z 0 [Note] [MY-000000] [Galera] async IST sender starting to serve tcp://127.0.0.1:4004 sending 21-21, preload starts from 21
2025-02-12T13:47:00.653193Z 1 [ERROR] [MY-000000] [WSREP] Invalid sst_request: clone<nullptr>clone_sst:47aedb42ae36592bc14d33da451a8c0dAa@127.0.0.1:4002<nullptr>
2025-02-12T13:47:00.653201Z 1 [Warning] [MY-000000] [Galera] SST preparation failed
```

kamil-holubicki replied:
fixed

venkatesh-prasad-v commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -0,0 +1,1251 @@
+MYPID=$$
+
+if [ "$WSREP_SST_OPT_PARENT" == "" ]; then
+        PARENT_PID=`cat /proc/$MYPID/status | grep PPid`
```
On my system, the below query did not return the PPID. Instead it returned `PPid:	369643`
```
$ cat /proc/557501/status | grep PPid
PPid:	369643
```
So, we must use fetch the PID using
```suggestion
        PARENT_PID=`cat /proc/$MYPID/status | grep PPid | awk '{print $2}`
```

kamil-holubicki replied:
@venkatesh-prasad-v What if it returns pid? Then print $2 will not work

kamil-holubicki replied:
@Tusamarco Why do we need this part at all? parent pid is passed as the parameter to the script
`--parent '160064'` on joiner side.
And we don't need it on Donor side.
I see this code is in common section. Should be only in joiner's section

kamil-holubicki replied:
removed this part from common. Only Joiner needs it

kamil-holubicki committed:
1. smart cleanup of Joiner's datadir
2. cleanup of parent process pid handling. It is needed only on Joiner side

kamil-holubicki committed:
Merge pull request #2032 from kamil-holubicki/address-review-comments-1

Address review comments 1

kamil-holubicki committed:
Mask password in error message.

kamil-holubicki committed:
Corner case:
Approach with 'ps' will cause problems if ps trucates executable path
because it is long (in my env it is truncated to 33 characters).
In such a case we will be looping in a 'while' loop. At some point
we will get to the directory level like /home or even / and during the
'find' we will error out with permissions error which will terminate
the script.

Fortunately we already have much better way to determine mysqld path.
It is 'get_mysqld_path' from wsrep_sst_common.sh. It also solves
problems with mysqldpath during Valgrind tests. Let's use it.

Tusamarco committed:
Merge pull request #2033 from kamil-holubicki/address-review-comments-2

Review comments addressed - part 2

kamil-holubicki committed:
PXC-4622:Joining node with clone SST method fails with component_keyring_file based encryption

https://perconadev.atlassian.net/browse/PXC-4622

Problem:
During SST we initialize a fresh directory for Joiner's clone instance.
During this initialization we use Joiner's original my.cnf. If this file
contains any encryption option enabled that expects keyring component
to be loaded the initialization fails if the server is configured to use
local manifest/keyring component config.

Cause:
This is because during the initialization the server tries to use
keyring component, as it is not loaded, initialization fails.
We can't place manifest/config in target directory, because
initialization expects empty data directory.

Solution:
For initialization time disable all encryption-related options that
can be reenabled after initialization.

There is only one option, innodb_sys_tablespace_encrypt, that requires
to be enabled during the intialization if we need encrypted system
tablespace. So this is incompatible with local manifest/config and we
can't disable it temporariliy during the initialization.
Detect such a case print error and terminate SST.

Additionally:
1. Moved get_json_value to wsrep_sst_common.sh
2. Introduced helper get_keyring_manifest_and_config_types function. Note
that it is similar to already existing get_keyring_manifest_and_config.
Consider refactoring.
3. Fixed hardcodded manifest name in wsrep_sst_xtrabackup-v2.sh (spotted
during the investigation)

Tusamarco committed:
Merge pull request #2034 from kamil-holubicki/PXC-4622

PXC-4622:Joining node with clone SST method fails with component_keyrng_file based encryption

Tusamarco committed:
Modify the SSL certification assignment and cleanup
Donor doesn't need any clone_ssl but it uses the ssl_% server section
While joiner requires the clone_ssl_%

kamil-holubicki committed:
Merge branch '8.0' into PXC-4469-merge-8.0

# Conflicts:
#	scripts/mysql_system_tables_fix.sql
#	scripts/mysql_system_users.sql

Tusamarco committed:
Merge pull request #2036 from kamil-holubicki/PXC-4469-merge-8.0

Merge branch 8.0 to PXC-4469

Tusamarco committed:
Merge branch 'PXC-4469-percona' into PXC-4469-fix-clonesst

* PXC-4469-percona:
  PXC-4573: SST Role lacks INNODB_REDO_LOG_ARCHIVE
  PXC-4498: PXC node "stalls" and causes the cluster to stop responding
  clang-tidy check fixed. 1. Boost url updated 2. 8.4 branch included

kamil-holubicki committed:
MTR tests fixed

Tusamarco reviewed (approved):
lgtm

Tusamarco committed:
Modify the SSL certification assignment and cleanup
Donor doesn't need any clone_ssl but it uses the ssl_% server section
While joiner requires the clone_ssl_%
Refactoring after discussion with Kamil

Tusamarco committed:
too fast in committing the changes

Tusamarco committed:
Cleanup for SSL configuration splitting the assignments by role Donor/Joiner

Tusamarco committed:
Cleanup for SSL configuration splitting the assignments by role Donor/Joiner Part 2 (with more isolation in the assignments)

Tusamarco committed:
space

Tusamarco committed:
space and debug

kamil-holubicki committed:
Merge pull request #2035 from Tusamarco/PXC-4469-fix-clonesst

Modify the SSL certification assignment and cleanup

kamil-holubicki committed:
1. Cleanup
2. Installed donor_cleanup handler in sst script
3. Fixed removing sst user on donor side. It should be dropped after
   sst script finish
4. Removed magic 'sleep 2'

kamil-holubicki committed:
Merge pull request #2040 from kamil-holubicki/PXC-4469-cleanup

PXC-4469 cleanup

Commit:
PXC-4469: Implement clone support for SSTs

***
Made the default of wsrep_sst_allowed_methods="xtrabackup-v2,clone"
to make server accept clone SST by default.

***
Test fixes due to the new default

Commit:
Merge pull request #2045 from venkatesh-prasad-v/add-clone-to-allowed-methods

PXC-4469: Implement clone support for SSTs

venkatesh-prasad-v reviewed (approved):
LGTM, approved!

kamil-holubicki commented:
Squashed and merged via https://github.com/percona/percona-xtradb-cluster/pull/2046
